jupyterlab_gcedetails/src/service/notebook_instance_service_layer.ts

```diff
@@ -0,0 +1,453 @@
+  enabled: boolean;
+}
+
+export const NOTEBOOKS_API_SERVICE: Readonly<Service> = {
+  name: 'AI Platform Notebooks API',
+  endpoint: 'notebooks.googleapis.com',
+  documentation: 'https://cloud.google.com/ai-platform/notebooks/',
+  isOptional: false,
+};
+
+type Operation = gapi.client.servicemanagement.Operation;
+type ListServicesResponse = gapi.client.servicemanagement.ListServicesResponse;
+
+const SERVICE_MANAGER = 'https://servicemanagement.googleapis.com/v1';
+export const NOTEBOOKS_API_PATH = 'https://notebooks.googleapis.com/v1beta1';
+
+const GOOGLE_API_UNAUTHENITCATED_CODE = 16;
+
+export function isUnauthorized(err: any) {
+  // Check for Google API Error structure and Unauthenticated error code
+  // https://cloud.google.com/apis/design/errors#error_codes
+  if (err.result && err.result.error) {
+    return err.result.error.code === GOOGLE_API_UNAUTHENITCATED_CODE;
+  }
+
+  return false;
+}
+
+/**
+ * Wraps around the GCP AI Platform Notebooks API to manage notebook resources in Google Cloud.
+ * https://cloud.google.com/ai-platform/notebooks/docs/reference/rest
+ */
+export class NotebookInstanceServiceLayer {
+  private projectIdPromise?: Promise<string>;
+  private instanceNamePromise?: Promise<string>;
+  private locationIdPromise?: Promise<string>;
+  private metadataPromise?: Promise<InstanceMetadata>;
+
+  constructor(
+    private _transportService: TransportService,
+    private _authToken: string,
+    projectId?: string,
+    instanceName?: string,
+    locationId?: string
+  ) {
+    if (projectId) {
+      this.projectIdPromise = Promise.resolve(projectId);
+    }
+    if (instanceName) {
+      this.instanceNamePromise = Promise.resolve(instanceName);
+    }
+    if (locationId) {
+      this.locationIdPromise = Promise.resolve(locationId);
+    }
+  }
+
+  get projectId() {
+    return this._getProject();
+  }
+
+  set projectId(projectIdPromise: string | Promise<string>) {
+    if (typeof projectIdPromise === 'string') {
+      this.projectIdPromise = Promise.resolve(projectIdPromise);
+    } else {
+      this.projectIdPromise = projectIdPromise;
+    }
+  }
+
+  get instanceName() {
+    return this._getInstance();
+  }
+
+  set instanceName(instanceNamePromise: string | Promise<string>) {
+    if (typeof instanceNamePromise === 'string') {
+      this.instanceNamePromise = Promise.resolve(instanceNamePromise);
+    } else {
+      this.instanceNamePromise = instanceNamePromise;
+    }
+  }
+
+  get locationId() {
+    return this._getLocation();
+  }
+
+  set locationId(locationIdPromise: string | Promise<string>) {
+    if (typeof locationIdPromise === 'string') {
+      this.locationIdPromise = Promise.resolve(locationIdPromise);
+    } else {
+      this.locationIdPromise = locationIdPromise;
+    }
+  }
+
+  set transportService(transportService: TransportService) {
+    this._transportService = transportService;
+  }
+
+  set authToken(authToken: string) {
+    this._authToken = authToken;
+  }
+
+  /**
```

Comment:
We can make the assumption that the Notebooks API will always be enabled for a user, unlike Scheduler, which relied on other services that are not directly related to CAIP Notebooks.